python/CMakeLists.txt

```diff
@@ -21,6 +21,8 @@
 cmake_minimum_required(VERSION 3.16)
 project(pyarrow)
 
+set(CMAKE_NO_SYSTEM_FROM_IMPORTED ON)
```

Comment:
Okay, this was a hard one.

CMake used to add Python include directories with `-isystem`, which led to some Python-internal includes to resolve to normal 3.13 includes, instead of 3.13-free-threading, which in turn meants that `Py_GIL_DISABLED` was not set.

Setting this flag uses `-I` instead. I verified manually that the only include directories here are python-specific (Python & NumPy include directories), so this shouldn't change too much. I couldn't find how I can change this in a more granular way. If someone knows that, help would be really appreaciated!

Reply:
@kou Does this change look ok to you?

Reply:
Can we use the `SYSTEM` target property https://cmake.org/cmake/help/latest/prop_tgt/SYSTEM.html instead of this to limit the impact?

```diff
diff --git a/python/CMakeLists.txt b/python/CMakeLists.txt
index 5d5eeaf815..c19820c074 100644
--- a/python/CMakeLists.txt
+++ b/python/CMakeLists.txt
@@ -258,6 +258,8 @@ set(EXECUTABLE_OUTPUT_PATH "${BUILD_OUTPUT_ROOT_DIRECTORY}")
 find_package(Python3Alt REQUIRED)
 message(STATUS "Found NumPy version: ${Python3_NumPy_VERSION}")
 message(STATUS "NumPy include dir: ${NUMPY_INCLUDE_DIRS}")
+# TODO: Describe why we need this
+set_target_properties(Python3::Python PROPERTIES SYSTEM FALSE)
 
 include(UseCython)
 message(STATUS "Found Cython version: ${CYTHON_VERSION}")
```

Reply:
> Can we use the `SYSTEM` target property https://cmake.org/cmake/help/latest/prop_tgt/SYSTEM.html instead of this to limit the impact?

What my change is the exact opposite though. It's signifying to not use `SYSTEM` anywhere, which I guess is what `FindPython` does.

Reply:
> > Can we use the `SYSTEM` target property https://cmake.org/cmake/help/latest/prop_tgt/SYSTEM.html instead of this to limit the impact?
> 
> What my change is the exact opposite though. It's signifying to not use `SYSTEM` anywhere, which I guess is what `FindPython` does.

Unless I am mistaken @kou is suggesting, to set `SYSTEM` to `FALSE` for Python but leave it as is for the rest of dependencies instead of changing it globally.

Reply:
I tested this, and, unfortunately it does not work properly.

Reply:
Like @lysnikolaou I tried several variations on this and I could not make it work. 

(such as `set_target_properties(Python3::Module PROPERTIES SYSTEM FALSE)`)

Reply:
The CMake docs are actually quite cryptic about this as several properties may be involved: `SYSTEM`, `NO_SYSTEM_FROM_IMPORTED` and `EXPORT_NO_SYSTEM`.

Reply:
@kou We'll have to live with this, unless you want to diagnose the issue yourself. Understanding CMake's intricacies is no fun.

Reply:
I can take a look at this, the system behavior is more compiler defined and exposed by CMake, -I always comes before any -isystem, so if you want to be certain you get it before system headers you need to force it. That is coupled with the other behavior that -isystem silences compiler warnings from system headers. It does seem a shame to change it globally, I see this was already merged but I will see if I can spot where it falls down being more surgical about it. In general -isystem is desirable for imported targets but it gets tricky when you have multiple versions of the same package floating around!

Reply:
OK. I'll try it later by myself too.